src/emulator/storage/apis/gcloud.ts

```diff
@@ -301,21 +354,49 @@ function sendFileBytes(
   res.setHeader("Accept-Ranges", "bytes");
   res.setHeader("Content-Type", md.contentType);
   res.setHeader("Content-Disposition", md.contentDisposition);
-  res.setHeader("Content-Encoding", "identity");
-
-  const byteRange = [...(req.header("range") || "").split("bytes="), "", ""];
-
-  const [rangeStart, rangeEnd] = byteRange[1].split("-");
-
-  if (rangeStart) {
-    const range = {
-      start: parseInt(rangeStart),
-      end: rangeEnd ? parseInt(rangeEnd) : data.byteLength,
-    };
-    res.setHeader("Content-Range", `bytes ${range.start}-${range.end - 1}/${data.byteLength}`);
-    res.status(206).end(data.slice(range.start, range.end));
+  res.setHeader("Content-Encoding", md.contentEncoding);
+  res.setHeader("ETag", md.etag);
+  res.setHeader("Cache-Control", md.cacheControl);
+  res.setHeader("x-goog-generation", `${md.generation}`);
+  res.setHeader("x-goog-metadatageneration", `${md.metageneration}`);
+  res.setHeader("x-goog-storage-class", md.storageClass);
+  res.setHeader("x-goog-hash", `crc32c=${crc32cToString(md.crc32c)},md5=${md.md5Hash}`);
+
+  const byteRange = req.range(data.byteLength, { combine: true });
+
+  if (Array.isArray(byteRange) && byteRange.type === "bytes" && byteRange.length > 0) {
+    const range = byteRange[0];
+    res.setHeader(
+      "Content-Range",
+      `${byteRange.type} ${range.start}-${range.end}/${data.byteLength}`
+    );
+    // Byte range requests are inclusive for start and end
+    res.status(206).end(data.slice(range.start, range.end + 1));
   } else {
     res.end(data);
   }
   return;
 }
+
+/** Sends 404 matching API */
+function sendObjectNotFound(req: Request, res: Response): void {
+  res.status(404);
+  const message = `No such object: ${req.params.bucketId}/${req.params.objectId}`;
+  if (req.method === "GET" && req.query.alt == "media") {
+    res.send(message);
+  } else {
+    res.json({
+      error: {
+        code: 404,
+        message,
+        errors: [
+          {
+            message,
+            domain: "global",
+            reason: "notFound",
+          },
+        ],
+      },
+    });
+  }
+}
```

Comment:
Tested with (`GET` with `alt=media`):

```
curl -i 'https://storage.googleapis.com/storage/v1/b/bookcreator-dev-firetool-tests/o/blah?alt=media'

HTTP/2 404 
x-guploader-uploadid: ADPycdst7ZSZzN7H2Yw38zD4PTx9Zu6wMutI4cgUbUWHUS366qbxzy8tklhFeSs9HeMXJj9-JWuFLhxj8cQ9X8GT88LqNALIfA
content-type: text/html; charset=UTF-8
date: Mon, 09 Aug 2021 12:22:42 GMT
vary: Origin
vary: X-Origin
expires: Mon, 09 Aug 2021 12:22:42 GMT
cache-control: private, max-age=0
content-length: 51
server: UploadServer
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"

No such object: bookcreator-dev-firetool-tests/blah
```

Tested with (`GET` without `alt=media`):

```
curl -i 'https://storage.googleapis.com/storage/v1/b/bookcreator-dev-firetool-tests/o/blah'

HTTP/2 404 
x-guploader-uploadid: ADPycdvIQMxtmw4ACLIf5rrG3GAMTQY7ST_w_BAZtGCGGke00EPB1bca_SAuLCeTYG9UeDx_625W-N78NwaiUtKPIjtWPU_mdw
content-type: application/json; charset=UTF-8
date: Mon, 09 Aug 2021 12:23:29 GMT
vary: Origin
vary: X-Origin
cache-control: no-cache, no-store, max-age=0, must-revalidate
expires: Mon, 01 Jan 1990 00:00:00 GMT
pragma: no-cache
content-length: 277
server: UploadServer
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"

{
  "error": {
    "code": 404,
    "message": "No such object: bookcreator-dev-firetool-tests/blah",
    "errors": [
      {
        "message": "No such object: bookcreator-dev-firetool-tests/blah",
        "domain": "global",
        "reason": "notFound"
      }
    ]
  }
}
```

And (non-`GET`):

```
curl -i 'https://storage.googleapis.com/storage/v1/b/bookcreator-dev-firetool-tests/o/blah' -X DELETE

HTTP/2 404 
x-guploader-uploadid: ADPycdt_7t20l4J1rLO0hosF99ICr446vfIK8TBzbt0cf9nQ_eO-0Rlsc_pegIV8VzcCJd3KcYgco2gQjKnOjrzk9P1vN0PN8A
content-type: application/json; charset=UTF-8
date: Mon, 09 Aug 2021 12:22:57 GMT
vary: Origin
vary: X-Origin
cache-control: no-cache, no-store, max-age=0, must-revalidate
expires: Mon, 01 Jan 1990 00:00:00 GMT
pragma: no-cache
content-length: 277
server: UploadServer
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"

{
  "error": {
    "code": 404,
    "message": "No such object: bookcreator-dev-firetool-tests/blah",
    "errors": [
      {
        "message": "No such object: bookcreator-dev-firetool-tests/blah",
        "domain": "global",
        "reason": "notFound"
      }
    ]
  }
}
```